Autohook in libox + other stuff

corpserot opened the pull request:
- More detailed instructions/documentation
- autohook needs to be available for it to work, and fallback gracefully to the old lua-only `debug.sethook` implementation whenever it's NOT available.
- game devs and mod devs need to set `sandbox.autohook = true` to enable autohook in their coroutine sandboxes.
- Instructions for Windows and Linux compilation, if the vendored library fails/crashes.
- Clarify support (experimental and WIP)
- Deleted `sbz_logic_autohook`
- Switch to xmake as the build tool.
- Rename to `lib*.dll` and `lib*.so` as these are the correct convention for libraries made with MinGW and Linux respectively.
- safely load the C module as much as we can.
- autohook C module now has versioning. to load autohook, both C source hash and LuaJIT version must match or it will be unavailable.
- autohook hook default implementation respects `time_limit` and `hook_time` coroutine sandbox specifications.
- autohook hook executes `def.in_hook` instead of its default implementation if it is defined. `in hook` can issue a kill command through `error()` to abort the sandbox bypassing auto-yielding.
- autohook hook checks if it has escaped the sandbox, and do nothing if so. otherwise, executes normally.
- and lots more detail if you go commit by commit. i'd rather you check the whole PR however.

corpserot committed:
Autohook in libox + other stuff

- More detailed instructions/documentation
- autohook now being locked behind an additional
  player setting and game/mod developer's sandbox
  spec.
- Windows support, with a bit of caveat.
- Clarify support (experimental and WIP)
- Deleted sbz_logic_autohook

corpserot reviewed (commented):
~~need some feedback~~ all done. plus these are far too outdated now.

TheEt1234 commented on mods/libox/autohook/README.md:
```diff
@@ -0,0 +1,35 @@
+
+3. With luanti compiled, it's recommended you bundle your DLLs since right now it will only run in the Clang64 environment. Follow the [instructions to bundle the DLLs](https://docs.luanti.org/for-engine-devs/compiling/windows/#bundling-dlls) in the same document. You'll be using [rollerozxa's bundledll bash script](https://github.com/rollerozxa/msys2-bundledlls)
+
+4. Now, we are ready to compile the autohook C module. First, install `mingw32-make` by running `paccman -Syu mingw-w64-clang-x86_64-make`. Afterwards, build the C module by running `make clean && make`.
```
Definitely need to change this line, that's not a real package and you are assuming arch linux
![image](https://github.com/user-attachments/assets/c565bbe9-b8cf-4d5e-9bc1-d467640ff7ea)

TheEt1234 commented on mods/libox/autohook/makefile:
yeah sure

TheEt1234 commented on mods/libox/coroutine.lua:
```diff
@@ -9,7 +9,10 @@ api.settings = {
         interval = 60
-    }
+    },
+    autohook = {
```
I'd prefer instead for libox to be more "generic", so instead of this i'd want something like `hook_setter_function = debug.sethook`

TheEt1234 commented on mods/libox/coroutine.lua:
```diff
@@ -275,7 +284,17 @@ function api.run_sandbox(ID, value_passed)
+                    attach_autohook()
+                else
+                    core.log('warning', ([[Autohook failed! Creating coroutine sandbox %q without autohook...
+(Did you enable the libox.autohook.enabled setting?)]]):format(ID))
```
remove it, libox without autohook should be a perfectly fine configuration

TheEt1234 commented on mods/libox/autohook/README.md:
```diff
@@ -0,0 +1,35 @@
+4. Done!
+
+## Compile: Windows
+This is a bit complicated. There's a shortcut you can use where you use (and trust) [ACorp's/corpserot's luanti build](https://github.com/corpserot/luanti-win-clang64) as the `autohook.dll` in libox is linked with that build's LuaJIT.
```
i dont know, maybe it would be a good idea to figure out what makes certain builds of luanti incompatible between eachother

TheEt1234 replied:
## WAAIT I think i know autohook is loaded in a weird way (not using `require`)

Maybe thats the issuse?

TheEt1234 replied:
could attempt to use ffi to load it? i mean... anything but using `require` xD (i don't know how it searches)

corpserot committed:
improve build system for autohook

- switch to xmake instead of make
- further clarify README.md
- use prefix-suffix that xmake produces (more correct)
  lib*.so on linux, and lib*.dll from mingw toolchains

corpserot committed:
remove separate autohook enable settings for players

- autohook now requires 1 argument, which is the sandbox spec.
- libox is less noisy when autohook is unavailable
- update windows autohook.

corpserot committed:
Merge remote-tracking branch 'frog/master' into frog-master

corpserot committed:
minor rename

corpserot committed:
fix: libox must check autohook during runtime

corpserot committed:
safer load and use autohook C module

corpserot committed:
autohook C module versioning and respects sandbox specs

- introduces autohook C module API versioning + checks
- autohook C module now respects the given hook_time and time_limit
  arguments.

corpserot committed:
update autohook docs

corpserot committed:
oops, autohook C module version check fail should be error

corpserot reviewed (commented):
~~need feedback~~ all done

corpserot committed:
Merge branch 'master' into frog-master

corpserot committed:
update autohook C module checks, versioning, build, etc

- better xmake build script with options that actually works
- use autohook.c hash as the version check
- no longer vendoring C modules
- update docs

corpserot committed:
sandbox.in_hook now runs from autohook too

corpserot committed:
autuhook fixes

- split general utils out of autohook.c
- add debug build and logging
- use hash of all C source files instead of just one
- when loading autohook, check for matching luajit version
- fix broken sandbox kill command for def.in_hook support
- update autohook documentation

corpserot committed:
Merge remote-tracking branch 'frog/master' into frog-master

corpserot committed:
autohook no-ops when it escapes the sandbox

corpserot committed:
autohook is no longer experimental and WIP

corpserot committed:
minor fix gitignore

corpserot committed:
various minor autohook things

- update knowledge station docs
- apply the same isyieldable check and attempts to
  in_hook called by autohook
- clean up old autohook usage

corpserot committed:
give luacontroller sandboxes access to whether its status

TheEt1234 commented on .gitignore:
```diff
@@ -1,4 +1,7 @@
 coverage/*
 mods/sbz_debug
 test_formspec
-mods/sbz_logic_autohook/bin/autohook.so
```
why remove this line? should the compiled files be in the repo?

TheEt1234 commented on mods/libox/autohook/.gitignore:
```diff
@@ -0,0 +1,6 @@
+compile_commands.json
+autohook.h
+*.dll
+*.so
```
oh i see my bad xD.... wait how does .gitignore work

Does any gitignore in any directory act as a .gitignore

TheEt1234 commented on mods/libox/autohook/README.md:
```diff
@@ -0,0 +1,88 @@
+# Compile the autohook C module
+## Compile: Linux/Unix-like systems
+1. Get [xmake](https://xmake.io/) either through its installer script or through your package manager.
+2. Install LuaJIT. You may install LuaJIT using your system's package manager, or with [xrepo](https://xrepo.xmake.io/).
```
Could mention that you shouldn't have luajit that's 8 years out of date

TheEt1234 commented on mods/libox/autohook/README.md:
```diff
@@ -0,0 +1,88 @@
+
+2. Follow instructions in the official documentation for engine devs to [compile Luanti on Windows using MSYS2 + Clang64 environment](https://docs.luanti.org/for-engine-devs/compiling/windows/). Don't use the MSVC + vcpkg approach, it's not really well-maintained as other methods. (e.g. using a really, really years old LuaJIT)
+
+   **WARNING**: the typical size of the tools installed is ~2GB.
```
[Offtopic] eww

TheEt1234 commented on mods/libox/autohook/autohook.c:
```diff
@@ -0,0 +1,210 @@
+static double prev_time = 0;
+static double get_time_ms(void) {
+    // On Windows, clock_t is 32-bit, meaning after 24 days and 20 hours it
+    // becomes UB. So uh... don't run your server for longer than that. - ACorp
```
any way around this? or is this fine

TheEt1234 commented on mods/libox/autohook/autohook.c:
```diff
@@ -0,0 +1,210 @@
+    int level = -1;
+    bool inside_sandbox = false;
+    while (lua_getstack(L, ++level, &parent)) {
+        lua_getinfo(L, "S", &parent);
```
i think calling lua_getinfo EVERY TIME a hook runs is a horrible idea for performance
i would ideally want the hook function to be as empty and as fast as possible

TheEt1234 replied:
ok, i will do a benchmark then

TheEt1234 commented on mods/libox/autohook/helpers.lua:
```diff
@@ -0,0 +1,40 @@
+        cmd = 'cat ' ..all_files ..' | ' find_program('shasum') " -a 256 | awk '{ print $1 }'"
+    end
+
+    -- this is so fucking cursed, and without select() it's not possible to bring pcall back...
```
nitpick: probably shouldnt swear in the code, im sorry, but i dont know if that would require putting up the "bad language" warning in contentDB

TheEt1234 commented on mods/libox/coroutine.lua:
```diff
@@ -40,6 +40,22 @@ end
 do_the_settings_thing("libox", api.settings)
 
+local attach_autohook = libox_autohook_module and libox_autohook_module.autohook
+local function attach_hook(sandbox)
```
Could maybe allow the user to make their own `attach_hook` function

TheEt1234 reviewed (commented):
looks fine,  also i'm curious: 
Why did autohook escape the sandbox more than a regular hook? and what is all that C code for that sounds like a lot

corpserot committed:
build autohook only with rolling release LuaJIT

also no swearing! we are family friendly
developers, not some garish barbaric punks!

corpserot committed:
Merge remote-tracking branch 'frog/master' into frog-master

corpserot committed:
hopefully fix compiling autohook on frog's machine

TheEt1234 commented:
...... why is autohook faster
![image](https://github.com/user-attachments/assets/03f3dde2-ee22-4f39-82c0-b888fa8d5f98)

TheEt1234 commented on mods/libox/api.md:
```diff
@@ -22,9 +22,11 @@
-`libox.disabled = true/false` - use when you want to disable libox (e.g. when using jitprofiler) 
+`libox.disabled = true/false` - use when you want to disable libox (e.g. when using jitprofiler)
+
+`libox.has_autohook = true/false` - Is autohook available or not?
```
Could maybe expose this to the luacontroller environment - Maybe the libox environment

TheEt1234 commented on mods/libox/autohook/README.md:
````diff
@@ -0,0 +1,88 @@
+xmake config -m debug -c && xmake
+```
+
+## All diagnostics when building
````
could you also add here your magic `xmake config -c -vD` for debugging

TheEt1234 commented on mods/libox/autohook/helpers.lua:
```diff
@@ -0,0 +1,42 @@
+if core ~= nil then error('DO NOT RUN THIS FROM LUANTI') end
```
Could have something in the readme that these files are xmake files, not luanti game code - do that if you want to

Also btw, you can do `assert(core, "DO NOT RUN THIS FROM LUANTI")` instead if you want to - assert is cool
(but only if you want to)

TheEt1234 commented on mods/sbz_logic/env.lua:
```diff
@@ -380,6 +380,7 @@ function logic.get_env(initial_pos, initial_meta, id)
         pos = vector.copy(initial_pos),
         yield = coroutine.yield,
         wait_for_event_type = wait_for_event_type,
+        has_autohook = libox.has_autohook,
```
oh you did that, my bad my bad
But maybe this also belongs in the libox environment instead

TheEt1234 reviewed (commented):
BTW, are you planning to make a seperate PR in libox, or should i just clone the changes to libox in some way

corpserot committed:
fixes to autohook

- gut out some unnecessary stuff in autohook.c
- remove sandbox check entirely because it can't
  see past 12 call stacks
- silence GCC warnings about compiler flags
- make has_autohook available in coroutine sandboxes by default

TheEt1234 commented:
ok :+1:

TheEt1234 commented:
sure ok lgtm
